src/query/ts/m3db/encoded_unconsolidated_step_iterator.go

```diff
@@ -21,99 +21,150 @@
 	"github.com/m3db/m3/src/query/block"
 	"github.com/m3db/m3/src/query/storage"
 	xts "github.com/m3db/m3/src/query/ts"
+	"github.com/m3db/m3/src/query/ts/m3db/consolidators"
 )
 
 type encodedStepIterUnconsolidated struct {
-	mu          sync.RWMutex
-	lastBlock   bool
-	exhausted   bool
-	err         error
-	meta        block.Metadata
-	validIters  []bool
-	seriesMeta  []block.SeriesMeta
-	seriesIters []encoding.SeriesIterator
+	lastBlock      bool
+	idx            int
+	err            error
+	meta           block.Metadata
+	seriesMeta     []block.SeriesMeta
+	seriesIters    []encoding.SeriesIterator
+	expandedSeries [][]xts.Datapoints
+
+	stepTime    time.Time
+	lookback    time.Duration
+	accumulator *consolidators.StepLookbackAccumulator
+	seriesPeek  []peekValue
 }
 
-func (it *encodedStepIterUnconsolidated) Current() (
-	block.UnconsolidatedStep,
+func (b *encodedBlockUnconsolidated) StepIter() (
+	block.UnconsolidatedStepIter,
 	error,
 ) {
-	it.mu.RLock()
-	if it.exhausted {
-		it.mu.RUnlock()
-		return nil, fmt.Errorf("out of bounds")
-	}
+	cs := b.consolidation
+	accumulator := consolidators.NewStepLookbackAccumulator(
+		b.lookback,
+		cs.bounds.StepSize,
+		cs.currentTime,
+		len(b.seriesBlockIterators),
+	)
+
+	return &encodedStepIterUnconsolidated{
+		idx:            -1,
+		meta:           b.meta,
+		seriesMeta:     b.seriesMetas,
+		seriesIters:    b.seriesBlockIterators,
+		lastBlock:      b.lastBlock,
+		expandedSeries: make([][]xts.Datapoints, len(b.seriesBlockIterators)),
+
+		lookback:    b.lookback,
+		accumulator: accumulator,
+		seriesPeek:  make([]peekValue, len(b.seriesBlockIterators)),
+	}, nil
+}
 
-	if it.err != nil {
-		it.mu.RUnlock()
-		return nil, it.err
+func (it *encodedStepIterUnconsolidated) Current() block.UnconsolidatedStep {
+	points := it.accumulator.AccumulateAndMoveToNext()
+	return storage.NewUnconsolidatedStep(it.stepTime, points)
+}
+
+func (it *encodedStepIterUnconsolidated) nextForStep(
+	i int,
+	stepTime time.Time,
+) {
+	peek := it.seriesPeek[i]
+	if peek.finished {
+		// No next value in this iterator
+		return
 	}
 
-	var t time.Time
-	values := make([]xts.Datapoints, len(it.seriesIters))
-	for i, iter := range it.seriesIters {
-		// Skip this iterator if it's not valid
-		if !it.validIters[i] {
-			continue
+	if peek.started {
+		point := peek.point
+		if point.Timestamp.After(stepTime) {
+			// This point exists further than the current step
+			// There are next values, but current step is empty.
+			return
 		}
 
-		dp, _, _ := iter.Current()
-		if i == 0 {
-			t = dp.Timestamp
-		}
+		// clear peeked point.
+		it.seriesPeek[i].started = false
+		// Currently at a potentially viable data point.
+		// Record previously peeked value, and all potentially valid values.
+		it.accumulator.AddPointForIterator(point, i)
 
-		values[i] = []xts.Datapoint{{
-			Timestamp: dp.Timestamp,
-			Value:     dp.Value,
-		}}
+		// If at boundary, add the point as the current value.
```

Comment:
Where do you add the point?

Reply:
It's a bit weird, but the point has been added to the consolidator at this point and it's already at the boundary so there's no reason to keep looking in the iterator until `Next` is called again. I agree the comment here is a bit misleading/crap though, will fix it up